pkg/gatewaysecret/handler_test.go

```diff
@@ -0,0 +1,351 @@
+		t.Run(testcase.name, func(t *testing.T) {
+			got, err := gatewaysecret.GetValidLastModifiedAt(testcase.secret)
+			if (err != nil) != testcase.wantErr {
+				t.Errorf("GetValidLastModifiedAt() error = %v, wantErr %v", err, testcase.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, testcase.want) {
+				t.Errorf("GetValidLastModifiedAt() got = %v, want %v", got, testcase.want)
+			}
+		})
+	}
+}
+
+func TestGatewaySecretRequiresUpdate(t *testing.T) {
+	type args struct {
+		gwSecret *apicorev1.Secret
+		caCert   certmanagerv1.Certificate
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "gateway secret is newer than CA certificate",
+			args: args{
+				gwSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:10Z",
+						},
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: false,
+		},
+		{
+			name: "gateway secret is older than CA certificate",
+			args: args{
+				gwSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:00Z",
+						},
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "gateway secret is newer but has parsing error for lastModifiedAt",
+			args: args{
+				gwSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:00",
+						},
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: true,
+		},
+	}
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			if got := gatewaysecret.GatewaySecretRequiresUpdate(
+				testcase.args.gwSecret, testcase.args.caCert); got != testcase.want {
+				t.Errorf("GatewaySecretRequiresUpdate() = %v, want %v", got, testcase.want)
+			}
+		})
+	}
+}
+
+func TestCopyRootSecretDataIntoGatewaySecret(t *testing.T) {
+	t.Parallel()
+
+	// Current gateway secret
+	gwSecret := &apicorev1.Secret{
+		ObjectMeta: apimetav1.ObjectMeta{
+			Annotations: map[string]string{
+				"lastModifiedAt": "2024-11-01T00:00:00Z",
+			},
+		},
+		Data: map[string][]byte{
+			"tls.crt": []byte(oldTLSCertValue),
+			"tls.key": []byte(oldTLSKeyValue),
+			"ca.crt":  []byte(oldCACertValue),
+		},
+	}
+
+	// Newer than gateway secret
+	caCert := certmanagerv1.Certificate{
```

Comment:
I think the `CopyRootSecretDataIntoGatewaySecret` function should be testable without having to fiddle with timestamps. Please see my other comment.